packages/user/src/model.ts

```diff
@@ -0,0 +1,200 @@
+// Copyright (c) Jupyter Development Team.
+// Distributed under the terms of the Modified BSD License.
+
+import { IStateDB } from '@jupyterlab/statedb';
+import { ISignal, Signal } from '@lumino/signaling';
+import { JSONObject, ReadonlyPartialJSONObject, UUID } from '@lumino/coreutils';
+import * as env from 'lib0/environment';
+
+import { ICurrentUser, IUser, USER } from './tokens';
+import { getAnonymousUserName, getRandomColor } from './utils';
+
+/**
+ * Default user implementation.
+ */
+export class User implements ICurrentUser {
+  private _id: string;
+  private _name: string;
+  private _username: string;
+  private _color: string;
+  private _anonymous: boolean;
+  private _role: IUser.ROLE;
+  private _cursor?: IUser.Cursor;
+
+  private _isReady = false;
+  private _state: IStateDB;
+  private _ready = new Signal<User, boolean>(this);
+  private _changed = new Signal<User, void>(this);
+
+  /**
+   * Constructor of the User class.
+   *
+   * @argument state: IStateDB to store users information.
+   */
+  constructor(state: IStateDB) {
+    this._state = state;
+    this._fetchUser()
+      .then(() => {
+        this._isReady = true;
+        this._ready.emit(true);
+      })
+      .catch(e => console.error(e));
+  }
+
+  /**
+   * User's ID.
+   */
+  get id(): string {
```

Comment:
It is not clear what the difference is between `id` and `username`. Are both needed? Can we document them in the interface?

Reply:
The reason for adding the id is that at the moment, every user is anonymous, and we do not have a unique username or email. Having the id allows us to create a unique UUID for every user.

Reply:
Ahh, thanks for the clarification. Although doesn't this mean that `username` is not necessarily unique. Given how `username` is usually tied to Auth systems, I think that is an important constraint for us to have as the username may be used in other systems or extensions as a unique key. In my mind a username as the following properties: 1) it is unique, 2) is more human-readable than a uuid, but 3) possibly less human readable than a first + family name. What if we removed the `id` field and instead made sure that the username was unique even when anonymous.

Reply:
I guess we can remove `id` and create a UUID as a `username` for anonymous users. But, we will have to use the `givenName` and `familyName` when showing the collaborators on the screen.

I believe this is much better than the actual implementation. Thanks for the help, I'll implement it.